text/3189-take-on-bool.md

```diff
@@ -0,0 +1,130 @@
+
+# Summary
+
+This RFC proposes adding a method to `bool` called `take`, which will save the value of the bool to a variable, set the bool to false, and then return that variable.
+
+# Motivation
+
+In many applications, deferred execution of routines is preferable, especially around long standing state. Using flags, or booleans which indicate a condition, is a good way to raise those routines. Taking that flag, while also resetting it, makes this pattern more elegant, without complexifying the `std` inordinately.
+
+The current one liner to get this same effect is `mem::take`. Adding `mem` into a project often implies a more complex operation than `.take` on a `Copy` tends to be, since `mem::take` (and `mem::replace` in general) is most useful when the type isn't Copy like `bool` is, which allows users to trivially do operations which otherwise require some non-trivial unsafe code. Moreover, in general, the `mem` module has a collection of relatively low-level operations on memory patterns. This makes it a more intimidating toolbox to reach for than a simple method on `bool` could be.
+
+There are two places where a "Take" pattern is used that is more about API usability, rather than memory handling, but where `mem::take` could be used: `Option` and `bool`. For the former, we already have the `.take` method (and in my Rust experience, anecdotally, is used often). We don't have anything for the latter. This PR's motivation is adding such a method for `bool`.
+
+# Guide-level explanation
+
+All .take() does is return the value of a boolean while also setting that value internally to false. It is just like `mem::take`, except it is called as a method instead of a free function. In places where booleans are commonly read and then reset, like dirty flags, this method is useful.
+
+For example, imagine a common game structure:
+
+```rs
+/// A recursive data structure of positions, commonly used in games to allow a parent/child relationship between transforms.
+pub struct SceneNode {
+    /// a position relative to this Node's parent.
+    pub local_position: [f32; 2],
+    /// an authoritative position
+    pub world_position: [f32; 2],
+
+    pub dirty: bool,
+}
+
+impl SceneNode {
+    /// We want a new local position!
+    pub fn set_local_position(&mut self, new_pos: [f32; 2]) {
+        self.local_position = new_pos;
+        self.dirty = true;
+    }
+
+    /// we magically have the parent in this example.
+    pub fn calculate_world_position(&mut self, parent: &SceneNode) {
+        // we can take the flag and also unset it in one method call
+        if self.dirty.take() {
+            self.world_position = [
+                self.local_position[0] + parent.local_position[0],
+                self.local_position[1] + parent.local_position[1],
+            ];
+        }
+
+        /*
+        // without this RFC, our code would be slightly uglier like this:
+        if self.dirty {
+            self.dirty = false;
```

Comment:
As a reader, I prefer this version. It is only one line longer and it is much clearer what is going on. `take` is not immediately obvious unless you are very familiar with the option version, and even then I think you might stop for a minute to consider the semantics (looking just at `self.dirty.take()` I might assume `dirty` is being treated as some kind of pointer and left as `null`).

(I also have a preference against functions which have both an important side effect and return an important value, they tend to be 'too clever' and to surprise readers).